src/Checkout/ExpressCheckout/Service/ExpressShippingCallbackService.php

```diff
@@ -8,134 +8,88 @@
-        );
-        $purchaseUnit->setAmount($amount);
+        $order = $this->orderBuilder->getOrderFromCart($cart, $salesChannelContext, new RequestDataBag());
+        $order->setId($callback->getId());
+        $order->getPurchaseUnits()->first()?->setReferenceId((string) $callback->getPurchaseUnits()->first()?->getReferenceId());
+        $order->getPurchaseUnits()->first()?->setShippingOptions($this->shippingOptionsProvider->getShippingOptions($salesChannelContext, $cart));
 
-        $purchaseUnitArray = $purchaseUnit->jsonSerialize();
+        $this->logger->debug('Shipping callback: cart recalculated', ['order' => $order]);
 
-        $this->logger->debug('Cart recalculation completed', [
-            'paypalOrderId' => $paypalOrderId,
-            'newCountry' => $countryCode,
-            'newTotal' => $amount->getValue(),
-            'newTax' => $amount->getBreakdown()?->getTaxTotal()?->getValue(),
-            'newShippingCosts' => $cart->getShippingCosts()->getTotalPrice(),
-        ]);
-
-        return [$purchaseUnitArray];
+        return $order;
     }
 
-    private function getCountryByIso(string $iso, Context $context): ?CountryEntity
+    private function switchSalesChannelContext(OrderShippingCallback $callback, SalesChannelContext $salesChannelContext): SalesChannelContext
     {
         $criteria = new Criteria();
-        $criteria->addFilter(new EqualsFilter('iso', $iso));
+        $criteria->addFilter(new EqualsFilter('iso', $callback->getShippingAddress()->getCountryCode()));
         $criteria->setLimit(1);
+        $criteria->getAssociation('states')
+            ->addFilter(new EqualsFilter('name', $callback->getShippingAddress()->getAdminArea1()))
+            ->setLimit(1);
+
+        $country = $this->countryRepository->search($criteria, $salesChannelContext->getContext())->getEntities()->first();
+        if (!$country) {
+            throw CheckoutException::expressCountryNotFound($callback->getShippingAddress()->getCountryCode());
+        }
+
+        // The current context should not be affected by the shipping country
+        // That the customer selects in the PayPal UI
+        // Hence a new context for the cart recalculation is created
+        $params = \array_filter([
+            SalesChannelContextService::COUNTRY_ID => $country->getId(),
+            SalesChannelContextService::COUNTRY_STATE_ID => $country->getStates()?->first()?->getId(),
+            SalesChannelContextService::SHIPPING_METHOD_ID => $callback->getShippingOption()?->getId(),
+        ]);
 
-        /** @var CountryEntity|null $country */
-        $country = $this->countryRepository->search($criteria, $context)->first();
+        $this->logger->debug('Shipping callback: switching context to new country', ['context_parameters' => $params]);
+        $token = $this->contextSwitchRoute->switchContext(new RequestDataBag($params), $salesChannelContext)->getToken();
```

Comment:
should we maybe check, if this stuff is actually different than the current SalesChannelContext? If not, we can skip a recalculation.